File: apps/mobile/src/features/widgets/widgetData.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as WidgetBridge from 'widget-bridge';

/**
 * Widget data bridge.
 *
 * Snapshot + pending actions are mirrored to:
 *   - AsyncStorage (JS fallback, always available, useful in dev/web)
 *   - Native shared storage via `widget-bridge` local Expo module:
 *       iOS:     App Group UserDefaults(suiteName: "group.com.indrox.theprimeway")
 *       Android: SharedPreferences("widget-data", MODE_PRIVATE)
 *
 * The native widget extension (iOS) / widget provider (Android) read from the
 * native store. Without the native bridge loaded (e.g. Expo Go, web), writes
 * fall back to AsyncStorage only and widgets show placeholder data.
 */

export const APP_GROUP_ID = 'group.com.indrox.theprimeway';

export interface WidgetTask {
  id: string;
  title: string;
  priority: 'low' | 'medium' | 'high';
  completed: boolean;
  timeSlot?: string;
}

export interface WidgetHabit {
  id: string;
  name: string;
  completed: boolean;
  streak: number;
}

export interface WidgetSnapshot {
  updatedAt: string;
  tasks: WidgetTask[];
  taskCompletedCount: number;
  taskTotal: number;
  currentStreak: number;
  longestStreak: number;
  nextHabit: WidgetHabit | null;
  pendingHabits: WidgetHabit[];
}

export interface PendingAction {
  id: string;
  kind: 'completeHabit' | 'completeTask';
  targetId: string;
  timestamp: string;
}

const KEY_SNAPSHOT = 'widget:snapshot';
const KEY_PENDING_ACTIONS = 'widget:pendingActions';

export async function writeWidgetSnapshot(snapshot: WidgetSnapshot): Promise<void> {
  const json = JSON.stringify(snapshot);
  await AsyncStorage.setItem(KEY_SNAPSHOT, json);
  try {
    WidgetBridge.writeSnapshot(json);
  } catch {
    // Native module unavailable — JS fallback only
  }
}

export async function readWidgetSnapshot(): Promise<WidgetSnapshot | null> {
  const raw = await AsyncStorage.getItem(KEY_SNAPSHOT);
  if (!raw) return null;
  try {
    return JSON.parse(raw) as WidgetSnapshot;
  } catch {
    return null;
  }
}

export async function readPendingActions(): Promise<PendingAction[]> {
  // Prefer native store — that's where widget check-ins land.
  let raw: string | null = null;
  try {
    raw = WidgetBridge.readPendingActions();
  } catch {
    raw = null;
  }
  if (!raw || raw === '[]') {
    raw = await AsyncStorage.getItem(KEY_PENDING_ACTIONS);
  }
  if (!raw) return [];
  try {
    return JSON.parse(raw) as PendingAction[];
  } catch {
    return [];
  }
}

export async function clearPendingActions(): Promise<void> {
  await AsyncStorage.removeItem(KEY_PENDING_ACTIONS);
  try {
    WidgetBridge.clearPendingActions();
  } catch {
    // ignore
  }
}

export function reloadWidgets(): void {
  try {
    WidgetBridge.reloadWidgets();
  } catch {
    // ignore
  }
}

File: apps/mobile/targets/widget/WidgetData.swift
import Foundation
import WidgetKit

// MARK: - Shared data shape (must match widgetData.ts)

struct WidgetTask: Codable, Identifiable {
  let id: String
  let title: String
  let priority: String
  let completed: Bool
  let timeSlot: String?
}

struct WidgetHabit: Codable, Identifiable {
  let id: String
  let name: String
  let completed: Bool
  let streak: Int
}

struct WidgetSnapshot: Codable {
  let updatedAt: String
  let tasks: [WidgetTask]
  let taskCompletedCount: Int
  let taskTotal: Int
  let currentStreak: Int
  let longestStreak: Int
  let nextHabit: WidgetHabit?
  let pendingHabits: [WidgetHabit]

  static let placeholder = WidgetSnapshot(
    updatedAt: ISO8601DateFormatter().string(from: Date()),
    tasks: [
      WidgetTask(id: "1", title: "Review quarterly goals", priority: "high", completed: false, timeSlot: "09:00"),
      WidgetTask(id: "2", title: "Deep work: auth refactor", priority: "medium", completed: false, timeSlot: "11:00"),
      WidgetTask(id: "3", title: "Team standup", priority: "low", completed: true, timeSlot: "10:00"),
    ],
    taskCompletedCount: 1,
    taskTotal: 3,
    currentStreak: 12,
    longestStreak: 28,
    nextHabit: WidgetHabit(id: "h1", name: "Meditate 10 min", completed: false, streak: 12),
    pendingHabits: [
      WidgetHabit(id: "h1", name: "Meditate 10 min", completed: false, streak: 12),
      WidgetHabit(id: "h2", name: "Read 20 pages", completed: false, streak: 5),
    ]
  )
}

// MARK: - Storage (MMKV writes as JSON string into App Group UserDefaults via the MMKV bridge;
// we read it from the App Group UserDefaults directly)

enum WidgetStorage {
  static let appGroup = "group.com.indrox.theprimeway"
  private static let snapshotKey = "snapshot"
  private static let pendingActionsKey = "pendingActions"

  static func defaults() -> UserDefaults? {
    UserDefaults(suiteName: appGroup)
  }

  static func loadSnapshot() -> WidgetSnapshot {
    guard let defaults = defaults(),
          let raw = defaults.string(forKey: snapshotKey),
          let data = raw.data(using: .utf8) else {
      return .placeholder
    }
    do {
      return try JSONDecoder().decode(WidgetSnapshot.self, from: data)
    } catch {
      return .placeholder
    }
  }

  struct PendingAction: Codable {
    let id: String
    let kind: String
    let targetId: String
    let timestamp: String
  }

  static func appendPendingAction(_ action: PendingAction) {
    guard let defaults = defaults() else { return }
    var current: [PendingAction] = []
    if let raw = defaults.string(forKey: pendingActionsKey),
       let data = raw.data(using: .utf8),
       let decoded = try? JSONDecoder().decode([PendingAction].self, from: data) {
      current = decoded
    }
    current.append(action)
    if let data = try? JSONEncoder().encode(current),
       let str = String(data: data, encoding: .utf8) {
      defaults.set(str, forKey: pendingActionsKey)
    }
  }
}
